### PR TITLE
SW-2408 API Wrappers for nursery withdrawals

### PR DESCRIPTION
**Pull request by karthikbtf**
- listNurseryWithdrawals
- hasNurseryWithdrawals (to gate nav menu)
- getNurseryWithdrawal by id
- reassignments is a separate ticket

Example response for `listNurseryWithdrawals`
```
[
    {
        "id": "3",
        "withdrawnDate": "2022-11-01",
        "purpose": "Nursery Transfer",
        "facility_name": "yo",
        "destinationName": "test",
        "totalWithdrawn": "2",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "4",
        "withdrawnDate": "2022-11-21",
        "purpose": "Dead",
        "facility_name": "yo",
        "totalWithdrawn": "2",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Dodonaea viscosa"
        ]
    },
    {
        "id": "6",
        "withdrawnDate": "2022-11-21",
        "purpose": "Dead",
        "facility_name": "yo",
        "totalWithdrawn": "16",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "7",
        "withdrawnDate": "2022-11-21",
        "purpose": "Dead",
        "facility_name": "test",
        "totalWithdrawn": "2",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "8",
        "withdrawnDate": "2022-11-21",
        "purpose": "Dead",
        "facility_name": "yo",
        "totalWithdrawn": "10",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "9",
        "withdrawnDate": "2022-11-21",
        "purpose": "Other",
        "facility_name": "yo",
        "totalWithdrawn": "8",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "11",
        "delivery_id": "1",
        "withdrawnDate": "2022-11-21",
        "purpose": "Out Plant",
        "facility_name": "yo",
        "destinationName": "plantsVzombies",
        "plotNames": "PZ4-C25",
        "totalWithdrawn": "3",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "12",
        "delivery_id": "2",
        "withdrawnDate": "2022-11-21",
        "purpose": "Out Plant",
        "facility_name": "yo",
        "destinationName": "test",
        "totalWithdrawn": "1",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "13",
        "withdrawnDate": "2022-11-21",
        "purpose": "Nursery Transfer",
        "facility_name": "yo",
        "destinationName": "ggg",
        "totalWithdrawn": "0",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "14",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "test",
        "totalWithdrawn": "0",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "15",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "test",
        "totalWithdrawn": "20",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "16",
        "withdrawnDate": "2022-11-22",
        "purpose": "Dead",
        "facility_name": "ggg",
        "totalWithdrawn": "10",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "19",
        "delivery_id": "5",
        "withdrawnDate": "2022-11-22",
        "purpose": "Out Plant",
        "facility_name": "yo",
        "destinationName": "test",
        "totalWithdrawn": "13",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "test1"
        ]
    },
    {
        "id": "20",
        "withdrawnDate": "2022-11-22",
        "purpose": "Dead",
        "facility_name": "ggg",
        "totalWithdrawn": "10",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "21",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "test",
        "totalWithdrawn": "6",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "22",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "ggg",
        "totalWithdrawn": "20",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "ajslkdfjl;asjflk;asd;lkfjasdlk;fjs"
        ]
    },
    {
        "id": "23",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "ggg",
        "totalWithdrawn": "1",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "24",
        "withdrawnDate": "2022-11-22",
        "purpose": "Other",
        "facility_name": "test",
        "totalWithdrawn": "2",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "25",
        "delivery_id": "6",
        "withdrawnDate": "2022-11-23",
        "purpose": "Out Plant",
        "facility_name": "test",
        "destinationName": "plantsVzombies",
        "plotNames": "PZ3-B17",
        "totalWithdrawn": "1",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "26",
        "withdrawnDate": "2022-11-23",
        "purpose": "Other",
        "facility_name": "ggg",
        "totalWithdrawn": "2",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa",
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "27",
        "withdrawnDate": "2022-11-23",
        "purpose": "Other",
        "facility_name": "ggg",
        "totalWithdrawn": "4",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abelia graebneriana var. deutziifoliaada"
        ]
    },
    {
        "id": "28",
        "delivery_id": "7",
        "withdrawnDate": "2022-11-29",
        "purpose": "Out Plant",
        "facility_name": "ggg",
        "destinationName": "lufita",
        "plotNames": "Lufita_PZ1-I-17",
        "totalWithdrawn": "5",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Abarema asplenifoliaasdfasdasdfa"
        ]
    },
    {
        "id": "29",
        "delivery_id": "8",
        "withdrawnDate": "2022-11-29",
        "purpose": "Out Plant",
        "facility_name": "test",
        "destinationName": "lufita",
        "plotNames": "Lufita_PZ1-I-17",
        "totalWithdrawn": "258",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Monstera deliciosa"
        ]
    },
    {
        "id": "30",
        "delivery_id": "9",
        "withdrawnDate": "2022-11-29",
        "purpose": "Out Plant",
        "facility_name": "iii",
        "destinationName": "lufita",
        "plotNames": "Lufita_PZ1-I-17",
        "totalWithdrawn": "1200",
        "hasReassignments": "false",
        "speciesScientificNames": [
            "Argemone glauca"
        ]
    }
]
```

Example response for `getWithdrawal by id`
```
{
    "requestSucceeded": true,
    "withdrawal": {
        "batchWithdrawals": [
            {
                "batchId": 13,
                "germinatingQuantityWithdrawn": 0,
                "notReadyQuantityWithdrawn": 1,
                "readyQuantityWithdrawn": 1
            }
        ],
        "destinationFacilityId": 77,
        "facilityId": 76,
        "id": 3,
        "purpose": "Nursery Transfer",
        "withdrawnDate": "2022-11-01"
    },
    "batches": [
        {
            "accessionId": 79,
            "addedDate": "2022-10-27",
            "batchNumber": "22-2-013",
            "facilityId": 76,
            "germinatingQuantity": 5,
            "id": 13,
            "latestObservedTime": "2022-11-01T20:17:07Z",
            "notReadyQuantity": 0,
            "readyQuantity": 0,
            "speciesId": 4,
            "version": 5
        }
    ]
}
```

### CODE REVIEW

File: src/types/Device.ts [hunk 1]
```diff
@@ -22,6 +22,6 @@ export type DeviceTemplate = {
   protocol?: string;
   address?: string;
   port?: number;
-  settings?: { [key: string]: { [key: string]: unknown } };
+  settings?: { [key: string]: unknown };
```
Comment (karthikbtf) [inline src/types/Device.ts:25]: this is from changes in the BE (see generated-schema)

**Comment by karthikbtf** [comment]
@constanzauanini @charlie-tf feel free to merge this if I am not around when this is approved.
This will unblock the table and the details views.

File: src/api/tracking/withdrawals.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    prefix: 'nurseryWithdrawals',
+    fields: [
+      'id',
+      'delivery_id',
+      'withdrawnDate',
+      'purpose',
+      'facility_name',
+      'destinationName',
+      'plotNames',
+      'batchWithdrawals.batch_species_scientificName',
```
Comment (sgrimm) [inline src/api/tracking/withdrawals.ts:32]: Change this to `batchWithdrawals.batch.species_scientificName` and you won't have to do the deduping later.

Comment (sgrimm) [reply src/api/tracking/withdrawals.ts:32]: That's a period rather than an underscore after `batch`.

Comment (karthikbtf) [reply src/api/tracking/withdrawals.ts:32]: Oh wow, awesome, let me try that. Thanks!!!

Comment (sgrimm) [reply src/api/tracking/withdrawals.ts:32]: In a search field path, you can think of `.` as, "Return a list of child objects" and `_` as, "Return a separate copy of the parent object for each child value."

Comment (sgrimm) [reply src/api/tracking/withdrawals.ts:32]: Oh, wait, hang on, I misread what the deduping code was doing. I think you still need it and the field name you have here is the right one. Sorry for the false alarm.

Comment (karthikbtf) [reply src/api/tracking/withdrawals.ts:32]: Oh no worries, thanks for looking at the review, really appreciate that.

**Review by constanzauanini — APPROVED** [review]

**Pull request merged**